File: frontend/src/utils/wasm/fonts.js
// Liberation font delivery for in-browser compliant generation.
// The engine cannot download fonts in WASM (pdfa.go EnsureFontsAvailable
// rejects); instead JS fetches TTF bytes from /fonts/ (vendored, OFL) and
// registers each under its STANDARD PDF name, mirroring the server flow in
// RegisterLiberationFontsForPDFA (pdfa.go:421-450).

import { ensureGopdfsuitWasm, asUint8Array, cachedFetch, callWasmObject } from './core.js'
import { PDFA_FONT_MANIFEST } from './manifests.generated.js'

// Standard name -> file. Generated at build time by
// scripts/check-wasm-manifests.mjs from internal/pdf/font/pdfa.go
// (LiberationFontMapping + LiberationFontFiles); re-exported here for
// backwards compatibility. Do not hand-edit: the prebuild check fails the
// build on drift.
export { PDFA_FONT_MANIFEST }

const FONT_BASE_URL = `${import.meta.env.BASE_URL}fonts`

const CACHE_NAME = 'gopdfsuit-fonts-v1'

async function fetchFontBytes(file) {
  const url = `${FONT_BASE_URL}/${file}`
  try {
    return await cachedFetch(url, { cacheName: CACHE_NAME, as: 'bytes' })
  } catch (err) {
    if (err instanceof TypeError) {
      // caches/fetch unavailable (or offline with empty cache): direct fetch.
      const response = await fetch(url)
      if (!response.ok) throw new Error(`font fetch failed: ${file} (${response.status})`)
      return new Uint8Array(await response.arrayBuffer())
    }
    throw err
  }
}

/**
 * Ensure the 12 Liberation faces are registered for compliant generation.
 * Fetches each TTF once (Cache API) and registers it under its standard
 * name. Returns { registered, missing, fetched }.
 */
export async function ensurePDFAFonts() {
  await ensureGopdfsuitWasm()
  const status = callWasmObject('goEnsurePDFAFonts', [])
  const missing = new Set(status.missing || [])
  let fetched = 0
  for (const entry of PDFA_FONT_MANIFEST) {
    if (!missing.has(entry.name)) continue
    const bytes = await fetchFontBytes(entry.file)
    callWasmObject('goRegisterFont', [entry.name, bytes])
    fetched += 1
  }
  const after = callWasmObject('goEnsurePDFAFonts', [])
  return { registered: after.registered || [], missing: after.missing || [], fetched }
}

/**
 * Register a single user-supplied TTF/OTF under a display name (font upload
 * path). Works offline; the face joins the same registry generate reads.
 */
export async function registerFontLocal(name, bytes) {
  await ensureGopdfsuitWasm()
  return callWasmObject('goRegisterFont', [name, asUint8Array(bytes)])
}

File: cmd/wasm/fonts.go
//go:build js && wasm

package main

import (
	"syscall/js"

	"github.com/chinmay-sawant/gopdfsuit/v7/internal/pdf/font"
)

// pdfaStandardFaces mirrors font.LiberationFontMapping keys: the standard
// names the engine resolves before embedding. JS registers each TTF under
// its standard name so the generator picks up subsets with no download.
var pdfaStandardFaces = []string{
	"Helvetica", "Helvetica-Bold", "Helvetica-Oblique", "Helvetica-BoldOblique",
	"Times-Roman", "Times-Bold", "Times-Italic", "Times-BoldItalic",
	"Courier", "Courier-Bold", "Courier-Oblique", "Courier-BoldOblique",
}

func registerWasmFontBindings() {
	js.Global().Set("goRegisterFont", js.FuncOf(registerFont))
	js.Global().Set("goEnsurePDFAFonts", js.FuncOf(ensurePDFAFonts))
}

// registerFont(name, bytes) parses TTF/OTF bytes and registers them under a
// standard PDF name (e.g. "Helvetica"). Returns {registered:name} or the
// shared {code,message,error} envelope.
func registerFont(_ js.Value, args []js.Value) (result any) {
	defer func() {
		if r := recover(); r != nil {
			if err, ok := r.(error); ok {
				result = errResult(err)
			} else {
				result = errResult(invalidInput("RegisterFont", "registration failed"))
			}
		}
	}()
	const op = "RegisterFont"
	if len(args) < 2 {
		return errResult(invalidInput(op, "expected (name, bytes)"))
	}
	if args[0].Type() != js.TypeString {
		return errResult(invalidInput(op, "expected font name as a string"))
	}
	name := args[0].String()
	if name == "" {
		return errResult(invalidInput(op, "empty font name"))
	}
	data, err := copyBytes(args[1], "font bytes", op)
	if err != nil {
		return errResult(err)
	}
	if err := font.GetFontRegistry().RegisterFontFromData(name, data); err != nil {
		return errResult(err)
	}
	return map[string]any{"registered": name}
}

// ensurePDFAFonts() reports which of the 12 standard faces are registered
// for compliant generation: {registered:[...], missing:[...]}.
func ensurePDFAFonts(_ js.Value, args []js.Value) (result any) {
	defer func() {
		if r := recover(); r != nil {
			if err, ok := r.(error); ok {
				result = errResult(err)
			} else {
				result = errResult(invalidInput("EnsurePDFAFonts", "status check failed"))
			}
		}
	}()
	registry := font.GetFontRegistry()
	var registered, missing []string
	for _, name := range pdfaStandardFaces {
		if registry.HasFont(name) {
			registered = append(registered, name)
		} else {
			missing = append(missing, name)
		}
	}
	if registered == nil {
		registered = []string{}
	}
	if missing == nil {
		missing = []string{}
	}
	return valueToJS(map[string]any{"registered": registered, "missing": missing})
}
